### DISCUSSION TITLE
City / place search features in map mode

### DISCUSSION THREAD

**Discussion by kha84** [deckerst/aves#337] [Ideas]
Don't want to raise a feature request yet, as I raised quite a lot already and don't feel comfortable telling dear @deckerst what to do with our lovely Aves app  :) So far I just wanted to discuss this my proposal first. 

I'm keep bringing my photo library in order and I noticed there's a scenario I follow all the time:
- find a photo without GPS
- open a dialog to assign GPS on it
- scroll-zoom-scroll-zoom the map to the spot where this photo belongs

For some of the occasions, it helps just to copy coordinates from some existing photo (see #329) but that is not covering all use cases:
- the photo you might want to use as a source for GPS data might not be displayed, considering your current filters
- it might be displayed, but it will take you a while until you find it. Means more scrolling

I just realized, what might help here a lot - if the map was having a textual search for Cities / Places. In this case if I want to assign a geotag to a bunch of photos I've made let's say near Niagara Falls, I would be doing something like this:

1) select those photos
2) edit -> edit location -> click on map icon
3) in a map mode I click on search button (the one I'm dreaming of) and in a popup dialog I enter "Niagara Falls" and map should navigate me there

This new search button can be also handy while just viewing your photos on a map. Like if you're interested to see what photos you've made in a specific city or place, you can click on it and type its name. So the map will adjust the view to be centered at that place. This saves you a time for scrolling and zooming a map.

Provided the #333 bug with Google Maps, I'd rather suggest this "Map search" button to appear for OpenStreetMaps first, if it needs to be implemented separately for every map.

What do you think about it?

**Comment by deckerst** [comment]
Being able to search for a city or place name is a good feature, but the issue here is economics: as far as I know, there is no free & unlimited API to do geocoding (i.e. place name -> coordinates). If there is, please let me know and I'll implement it.

There is a subpar alternative, which would be to provide this field to search among the places already known from the user collection. This variant is possible because from the place name I can retrieve the matching photos, average their coordinates, and center the map on this. But I don't think this is a good idea in practice because users seeing the search field will expect more than searching among known places.

**Reply by kha84** [reply]
Yeap, absolutely agree. I'll have a look if there's such API

**Reply by kha84** [reply]
Does this one fit the purpose?

https://nominatim.org/release-docs/latest/api/Overview/

https://nominatim.openstreetmap.org/ui/search.html

**Reply by deckerst** [reply]
Nope, they (rightfully) request "No heavy uses":
https://operations.osmfoundation.org/policies/nominatim/

**Reply by kha84** [reply]
But the use will be moderate and requests will be coming distributed, from user devices, right? You're not proxying user requests via some central server. So if some abusing user will be blocked, well it's his own fault.
Or am I missing something?

**Reply by deckerst** [reply]
If I were them, I'd ban the "User-Agent" (which is also a sensible requirement, as it identifies the app/website) so no user of Aves, wherever they are, would be able to use the API.

As for "moderate usage", I don't adopt APIs that impose limitations, because I don't want to care about this later if the app gets more popular. Especially, Nominatim & OSM are founded by donations, and I definitely don't want to abuse their services.

**Reply by kha84** [reply]
Sure I get that. I'm guessing in order to use geo decoding features from Google you'll need to plug some API key into the app, which will be shared by all users, so again - can be banned at any moment.
Looks like the only feasible solution would be to have some small cache user can download if he wants to use this kind of searching feature :)

**Reply by deckerst** [reply]
I already use an API key for Google/Huawei services. Because this is my key, it's not in the public source but injected during the build process. If another developer wants to fork and release their version, they'll have to use their own key if they want to keep the Google/Huawei related features.

With the Google services, the reverse geocoding endpoint is unlimited (that's how you get the place and country names from the photo coordinates), but the geocoding endpoint is limited, so I'm not using it.

**Reply by kha84** [reply]
Well this feature request can be put to a backburner then. I'll try to see if they're any other geocoding services can be used, but I get the point - you already researched most of viable options.
